mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td

```diff
@@ -260,74 +256,380 @@ def MIOpen_WorkitemIdOp:
+    the change in the lower indices compared to `lowerOrig'.
+  }];
+
+  let builders = [
+    OpBuilder<(ins "TransformMapAttr":$transform, "ValueRange":$upperDiffs,
+      "ValueRange":$lowerOrig)>
+  ];
+
+  let assemblyFormat = [{
+    attr-dict $map `(` $upperDiffs `)` `+` `(` $lowerOrig `)` `:`
+      type($lowerIndices)
+  }];
+  let hasVerifier = 1;
+}
+
+// buffer_load
+def MIOpen_BufferLoadOp :
+    MIOpen_Op<"buffer_load">,
+    Arguments<(ins Arg<MemRefOf<[F32, F16, BF16, I8, I32]>,
+        "buffer to load from", [MemRead]>:$source,
+      BoolArrayAttr:$oobDims,
+      Variadic<Index>:$coords)>,
+    Results<(outs AnyTypeOf<[F32, F16, BF16, I8, I32,
+              VectorOfLengthAndType<[2, 4], [F32]>,
+              VectorOfLengthAndType<[2, 4, 8], [F16]>,
+              VectorOfLengthAndType<[2, 4, 8], [BF16]>,
+              VectorOfLengthAndType<[4], [I8]>,
+              VectorOfLengthAndType<[2, 4], [I32]>]>:$result)> {
+  let summary = "Load data from a global buffer";
+
+  let description = [{
+    miopen.buffer_load uses, if necessary, GPU buffer load intrinsics to
+    ensure that out of bounds loads return 0 instead of a garbage value.
+
+    The bounds on each coordinate are the size of the memref in that coordinate.
+
+    This op can perform vector reads.
+
+    The memref must be in global memory (memory space 0).
+
+    `oobDims` are the dimensions to be checked for out of bounds indices.
   }];
   let assemblyFormat = [{
-    $data `->` $dest `[` $destCoord `]` `with` $transforms attr-dict
-    `:` type($data) `->` type($dest) `,` type($destCoord)
+    $source `[` $coords `]` attr-dict
+    `:` type($source) `,` type($coords) `->` type($result)
   }];
+  let hasVerifier = 1;
 }
 
+// buffer_store
+def MIOpen_BufferStoreOp :
+    MIOpen_Op<"buffer_store", []>,
+    Arguments<(ins AnyTypeOf<[F32, F16, BF16, I8, I32,
+        VectorOfLengthAndType<[2, 4], [F32]>,
+        VectorOfLengthAndType<[2, 4, 8], [F16]>,
+        VectorOfLengthAndType<[2, 4, 8], [BF16]>,
+        VectorOfLengthAndType<[4], [I8]>,
+        VectorOfLengthAndType<[2, 4], [I32]>]>:$data,
+      Arg<MemRefOf<[F32, F16, BF16, I8, I32]>,
+        "Buffer to store to", [MemWrite]>:$dest,
+      BoolArrayAttr:$oobDims,
+      Variadic<Index>:$coords,
+      OptionalAttr<MIOpen_PaddingInfoAttr>:$paddingInfo,
+      OptionalAttr<InMemoryDataOperation>:$dataOperation)> {
+  let summary = "Store data to a global buffer";
+
+  let description = [{
+    `miopen.buffer_store` stores data to a global buffer, ignoring out of bounds writes.
+
+    The bounds on a write are the dimensions of the memref, though only the
+    dimensions specified in `oobDims` are checked before the write is issued.
+
+    Writes that our out of bounds are ignored.
+
+    This op supports vector writes and uses buffer store intrinsics if needed.
+
+    `paddingInfo` is used to apply backwards data store hacks if they're needed.
+
+    The buffer must reside in global memory.
```

Comment:
Reading the code it seems the check sits in the verifier method.

Wondering is there a way in TableGen to guarantee that with a Trait? This can be done in the future PRs.

Reply:
I didn't see a trait for memory space, that does feel like a nice todo item though